File: packages/typescript/src/catalogs/ift/codigos-lada.ts
/**
 * Catálogo de códigos LADA (plan de numeración telefónica) en México (IFT)
 */

import * as path from 'path';
import * as fs from 'fs';
import { CodigoLADA } from '../../types';

interface CodigosLADAData {
  metadata: {
    catalog: string;
    description: string;
    source: string;
    last_updated: string;
    notes: string;
  };
  codigos: CodigoLADA[];
  tipos_lada: Record<string, string>;
  regiones: string[];
  marcacion_actual: {
    descripcion: string;
    formato: string;
    ejemplo: string;
    prefijos_obsoletos: string[];
    vigencia: string;
  };
  estadisticas: {
    total_codigos: number;
    codigos_metropolitanos: number;
    codigos_fronterizos: number;
    codigos_turisticos: number;
    estados_cubiertos: number;
  };
}

export class CodigosLADA {
  private static _data: CodigoLADA[] | null = null;

  private static loadData(): void {
    if (this._data !== null) return;

    const dataPath = path.resolve(__dirname, '../../../../shared-data/ift/codigos_lada.json');
    const rawData = fs.readFileSync(dataPath, 'utf-8');
    const jsonData: CodigosLADAData = JSON.parse(rawData);
    this._data = jsonData.codigos;
  }

  /**
   * Obtener todos los códigos LADA
   */
  static getAll(): CodigoLADA[] {
    this.loadData();
    return this._data!;
  }

  /**
   * Buscar código LADA específico
   */
  static buscarPorLADA(lada: string): CodigoLADA | undefined {
    this.loadData();
    return this._data!.find((c) => c.lada === lada);
  }

  /**
   * Buscar códigos por ciudad
   */
  static buscarPorCiudad(ciudad: string): CodigoLADA[] {
    this.loadData();
    return this._data!.filter((c) => c.ciudad.toLowerCase().includes(ciudad.toLowerCase()));
  }

  /**
   * Obtener códigos por estado
   */
  static getPorEstado(estado: string): CodigoLADA[] {
    this.loadData();
    return this._data!.filter((c) => c.estado.toLowerCase() === estado.toLowerCase());
  }

  /**
   * Obtener códigos por tipo
   */
  static getPorTipo(tipo: 'metropolitana' | 'fronteriza' | 'turistica' | 'normal'): CodigoLADA[] {
    this.loadData();
    return this._data!.filter((c) => c.tipo === tipo);
  }

  /**
   * Obtener códigos por región
   */
  static getPorRegion(region: string): CodigoLADA[] {
    this.loadData();
    return this._data!.filter((c) => c.region.toLowerCase() === region.toLowerCase());
  }

  /**
   * Obtener ciudades metropolitanas
   */
  static getMetropolitanas(): CodigoLADA[] {
    return this.getPorTipo('metropolitana');
  }

  /**
   * Obtener ciudades fronterizas
   */
  static getFronterizas(): CodigoLADA[] {
    return this.getPorTipo('fronteriza');
  }

  /**
   * Obtener destinos turísticos
   */
  static getTuristicas(): CodigoLADA[] {
    return this.getPorTipo('turistica');
  }

  /**
   * Validar formato de número telefónico (10 dígitos)
   */
  static validarNumero(numero: string): {
    valid: boolean;
    lada?: string;
    numero_local?: string;
    ciudad?: string;
    estado?: string;
    error?: string;
  } {
    // Eliminar espacios y guiones
    const numeroLimpio = numero.replace(/[\s-]/g, '');

    // Validar que sean 10 dígitos
    if (!/^\d{10}$/.test(numeroLimpio)) {
      return {
        valid: false,
        error: 'El número debe tener exactamente 10 dígitos',
      };
    }

    // Intentar extraer LADA (primeros 2 o 3 dígitos)
    this.loadData();

    // Intentar con 3 dígitos
    let lada = numeroLimpio.substring(0, 3);
    let codigo = this._data!.find((c) => c.lada === lada);

    // Si no se encuentra, intentar con 2 dígitos
    if (!codigo) {
      lada = numeroLimpio.substring(0, 2);
      codigo = this._data!.find((c) => c.lada === lada);
    }

    if (codigo) {
      const numeroLocal = numeroLimpio.substring(lada.length);
      return {
        valid: true,
        lada: codigo.lada,
        numero_local: numeroLocal,
        ciudad: codigo.ciudad,
        estado: codigo.estado,
      };
    }

    return {
      valid: false,
      lada: lada,
      error: `Código LADA ${lada} no encontrado en el catálogo`,
    };
  }

  /**
   * Formatear número telefónico
   */
  static formatearNumero(numero: string): string {
    const validacion = this.validarNumero(numero);
    if (!validacion.valid || !validacion.lada || !validacion.numero_local) {
      return numero;
    }

    // Formato: LADA XXXX XXXX
    const local = validacion.numero_local;
    if (local.length === 7) {
      return `${validacion.lada} ${local.substring(0, 3)} ${local.substring(3)}`;
    } else if (local.length === 8) {
      return `${validacion.lada} ${local.substring(0, 4)} ${local.substring(4)}`;
    }
    return `${validacion.lada} ${local}`;
  }

  /**
   * Obtener información de un número telefónico
   */
  static getInfoNumero(numero: string): {
    lada: string;
    ciudad: string;
    estado: string;
    tipo: string;
    region: string;
  } | null {
    const validacion = this.validarNumero(numero);
    if (!validacion.valid || !validacion.lada) {
      return null;
    }

    const codigo = this.buscarPorLADA(validacion.lada);
    if (!codigo) {
      return null;
    }

    return {
      lada: codigo.lada,
      ciudad: codigo.ciudad,
      estado: codigo.estado,
      tipo: codigo.tipo,
      region: codigo.region,
    };
  }
}

File: packages/dart/lib/src/catalogs/ift/codigos_lada.dart
/// Catálogo de Códigos LADA de México
///
/// Basado en el Plan de Numeración Nacional del IFT (Instituto Federal de Telecomunicaciones).
/// Incluye mapeo geográfico a códigos INEGI y zonas metropolitanas.
library;

import 'package:catalogmx/src/catalogs/base_catalog.dart';

/// Catálogo de códigos LADA con mapeo geográfico
class CodigosLADACatalog {
  static List<Map<String, dynamic>>? _data;
  static Map<String, Map<String, dynamic>>? _byLada;
  static Map<String, List<Map<String, dynamic>>>? _byEntidad;

  /// LADAs de 2 dígitos (principales ciudades metropolitanas)
  static const List<String> ladas2Digitos = ['55', '33', '81'];

  static void _loadData() {
    if (_data != null) return;

    final jsonData = BaseCatalog.loadJsonDataSync('ift/codigos_lada.json');

    // Handle the wrapped format with "codigos" key
    if (jsonData.isNotEmpty && jsonData.first.containsKey('codigos')) {
      final codigos = jsonData.first['codigos'];
      if (codigos is List) {
        _data = codigos.cast<Map<String, dynamic>>();
      } else {
        _data = jsonData;
      }
    } else {
      _data = jsonData;
    }

    // Build indexes
    _byLada = {for (var item in _data!) item['lada'] as String: item};

    _byEntidad = {};
    for (var item in _data!) {
      final ent = item['cve_entidad'] as String?;
      if (ent != null) {
        _byEntidad![ent] ??= [];
        _byEntidad![ent]!.add(item);
      }
    }
  }

  /// Obtiene todos los códigos LADA
  static List<Map<String, dynamic>> getAll() {
    _loadData();
    return List.from(_data!);
  }

  /// Busca información por código LADA
  static Map<String, dynamic>? buscarPorLada(String lada) {
    _loadData();
    return _byLada![lada];
  }

  /// Alias para buscarPorLada
  static Map<String, dynamic>? getByLada(String lada) => buscarPorLada(lada);

  /// Busca códigos LADA por nombre de ciudad
  static List<Map<String, dynamic>> buscarPorCiudad(String ciudad) {
    _loadData();
    final ciudadUpper = ciudad.toUpperCase();
    return _data!
        .where(
          (item) => (item['ciudad'] as String? ?? '')
              .toUpperCase()
              .contains(ciudadUpper),
        )
        .toList();
  }

  /// Obtiene códigos LADA por estado
  static List<Map<String, dynamic>> getPorEstado(String estado) {
    _loadData();
    final estadoUpper = estado.toUpperCase();
    return _data!
        .where(
          (item) => (item['estado'] as String? ?? '')
              .toUpperCase()
              .contains(estadoUpper),
        )
        .toList();
  }

  /// Obtiene códigos LADA por clave de entidad INEGI
  static List<Map<String, dynamic>> getPorCveEntidad(String cveEntidad) {
    _loadData();
    return List.from(_byEntidad![cveEntidad] ?? []);
  }

  /// Obtiene códigos LADA por tipo (metropolitana, fronteriza, turistica, normal)
  static List<Map<String, dynamic>> getPorTipo(String tipo) {
    _loadData();
    return _data!.where((item) => item['tipo'] == tipo).toList();
  }

  /// Obtiene códigos LADA por región
  static List<Map<String, dynamic>> getPorRegion(String region) {
    _loadData();
    return _data!.where((item) => item['region'] == region).toList();
  }

  /// Valida si una LADA existe
  static bool isValid(String lada) {
    _loadData();
    return _byLada!.containsKey(lada);
  }

  /// Valida un número de teléfono mexicano
  static Map<String, dynamic> validarNumero(String numero) {
    _loadData();

    // Clean number
    final numeroClean = numero.replaceAll(RegExp(r'[^\d]'), '');

    final result = <String, dynamic>{
      'numero': numeroClean,
      'valido': false,
      'lada': null,
      'ciudad': null,
      'estado': null,
      'tipo': null,
      'error': null,
    };

    if (numeroClean.length != 10) {
      result['error'] = 'Debe tener 10 dígitos';
      return result;
    }

    // Try 2-digit LADA first (55, 33, 81)
    final lada2 = numeroClean.substring(0, 2);
    if (ladas2Digitos.contains(lada2)) {
      final info = _byLada![lada2];
      if (info != null) {
        result['valido'] = true;
        result['lada'] = lada2;
        result['ciudad'] = info['ciudad'];
        result['estado'] = info['estado'];
        result['tipo'] = info['tipo'];
        return result;
      }
    }

    // Try 3-digit LADA
    final lada3 = numeroClean.substring(0, 3);
    final info = _byLada![lada3];
    if (info != null) {
      result['valido'] = true;
      result['lada'] = lada3;
      result['ciudad'] = info['ciudad'];
      result['estado'] = info['estado'];
      result['tipo'] = info['tipo'];
      return result;
    }

    result['error'] = 'LADA no reconocida';
    return result;
  }

  /// Formatea un número de teléfono mexicano
  static String formatearNumero(String numero) {
    final numeroClean = numero.replaceAll(RegExp(r'[^\d]'), '');

    if (numeroClean.length != 10) {
      return numero;
    }

    // Check if 2-digit LADA
    if (ladas2Digitos.contains(numeroClean.substring(0, 2))) {
      return '${numeroClean.substring(0, 2)} ${numeroClean.substring(2, 6)} ${numeroClean.substring(6)}';
    }

    // 3-digit LADA
    return '${numeroClean.substring(0, 3)} ${numeroClean.substring(3, 6)} ${numeroClean.substring(6)}';
  }

  /// Obtiene LADAs metropolitanas (principales ciudades)
  static List<Map<String, dynamic>> getMetropolitanas() {
    return getPorTipo('metropolitana');
  }

  /// Obtiene LADAs fronterizas
  static List<Map<String, dynamic>> getFronterizas() {
    return getPorTipo('fronteriza');
  }

  /// Obtiene LADAs turísticas
  static List<Map<String, dynamic>> getTuristicas() {
    return getPorTipo('turistica');
  }

  /// Obtiene estadísticas del catálogo
  static Map<String, dynamic> getEstadisticas() {
    _loadData();

    final tipos = <String, int>{};
    final regiones = <String, int>{};
    final estados = <String, int>{};

    for (var item in _data!) {
      final tipo = item['tipo'] as String? ?? 'unknown';
      tipos[tipo] = (tipos[tipo] ?? 0) + 1;

      final region = item['region'] as String? ?? 'unknown';
      regiones[region] = (regiones[region] ?? 0) + 1;

      final estado = item['estado'] as String? ?? 'unknown';
      estados[estado] = (estados[estado] ?? 0) + 1;
    }

    return {
      'total_codigos': _data!.length,
      'por_tipo': tipos,
      'por_region': regiones,
      'estados_cubiertos': estados.length,
    };
  }

  /// Limpia el cache
  static void clearCache() {
    _data = null;
    _byLada = null;
    _byEntidad = null;
  }
}
